src/backend/distributed/utils/metadata_cache.c

```diff
@@ -563,6 +563,53 @@ LookupNodeForGroup(uint32 groupId)
 	}
 }
 
+/*
+ * NodeShardList returns the list of shards that are hosted at the given node.
+ *
+ * Basically we're doing the following SQL query:
+ *
+    select shardid
+      from pg_dist_node
+           join pg_dist_placement using(groupid)
+    where nodeid = <nodeId>;
+ */
+List *
+NodePlacementList(int32 nodeId)
+{
+	int32 groupId = -1;
+
+	/*
+	 * First get the groupId for the given nodeId.
+	 */
+	PrepareWorkerNodeCache();
+
+	for (int i = 0; i < WorkerNodeCount; i++)
+	{
+		WorkerNode *workerNode = WorkerNodeArray[i];
+
+		if (workerNode->nodeId == nodeId)
+		{
+			groupId = workerNode->groupId;
+			break;
+		}
+	}
+
+	if (groupId == -1)
+	{
+		ereport(ERROR,
+				(errcode(ERRCODE_NO_DATA_FOUND),
+				 errmsg("node %u not found", nodeId)));
+
+		/* keep compiler happy */
+		return NIL;
+	}
+
+	/*
+	 * We now have the groupId for our nodeId and we can scan pg_dist_placement
+	 * for the list of shards found on the node.
+	 */
+	return AllShardPlacementsOnNodeGroup(groupId);
```

Comment:
minor: we usually don't do tail calls and rather have an intermediate variable such that you can easily put a breakpoint on the return or step through the function and inspect the result before returning.

Reply:
done